### PR TITLE
refactor: removed dynamic scoping from shared module

### LINKED ISSUE #4248: [refactor] remove dynamic scoping from shared module

### PR DESCRIPTION
**Pull request by alexandraB99**
Closes #4248

### What does the PR do
Properties and functions moved to Global:
- toastMessage
- openProfilePopup()
- getProfileImage()
- popupOpened

### Affected areas
All areas using shared components

### CODE REVIEW

File: ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml [hunk 1]
```diff
@@ -78,7 +78,7 @@ Popup {
             appMain.changeAppSection(Constants.profile)
             // TODO: replace with shared store constant
             // Profile/RootStore.notifications_id
-            profileLayoutContainer.changeProfileSection(7)
+            Global.changeProfileSection(7)
```
Comment (PascalPrecht) [inline ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:81]: This kinda feels weird. Why does the global scope have to know how to change profile sections?
Imagine a scenario where we're able lazy load modules, then the profile module may not even be instantiated yet, still there'd be an API to change profile sections.

I don't know from the top of my head what the function does, but judging by what it's called it feels like this wants to live in the profile module only.

File: ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml [hunk 2]
```diff
@@ -48,12 +48,12 @@ Popup {
             color: Style.current.dropShadow
         }
     }
-    x: applicationWindow.width - activityCenter.width - Style.current.halfPadding
+    x: Global.applicationWindow.width - activityCenter.width - Style.current.halfPadding
     onOpened: {
-        popupOpened = true
+        Global.popupOpened = true
```
Comment (Khushboo-dev-cpp) [inline ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:53]: this seems tricky, is this to know if any popup is launched any where?

Can we then not use the instance created in the AppMain.qml when calling openPopup() to connect to th onOpened and inclosed Signals and set it there? Would be become more generic then?

File: ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml (hunk 1 above)
Comment (Khushboo-dev-cpp) [inline ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:81]: cna this be combined with the changeAppSection API? as a param? Just a thought for cleaner API. then in case we need to handle a use case for any other section, we won't have to write new methods.

File: ui/app/AppLayouts/Chat/views/ActivityCenterMessageComponentView.qml [hunk 3]
```diff
@@ -253,7 +250,7 @@ Item {
 //                }
 //                return chatType
             }
-            profileImage: realChatType === Constants.chatType.oneToOne ? appMain.getProfileImage(chatId) || ""  : ""
+            profileImage: realChatType === Constants.chatTypeOneToOne ? Global.getProfileImage(chatId) || ""  : ""
```
Comment (PascalPrecht) [inline ui/app/AppLayouts/Chat/views/ActivityCenterMessageComponentView.qml:253]: Not sure if this is actually part of this refactor but `chatType.oneToOne` has more of a enum-ish kind of feel to it than `chatTypeOneToOne`.

If this was changed in this PR, any particular reasoning this was done?
(If it was not, ignore this question)

File: ui/app/AppLayouts/Chat/views/ChatTextView.qml [hunk 4]
```diff
@@ -94,12 +94,21 @@ Item {
...
 //                return;
 //            }
+=======
+            if (link.startsWith('//')) {
+                let pk = link.replace("//", "");
+                const userProfileImage = Global.getProfileImage(pk)
+                Global.openProfilePopup(root.store.userNameOrAlias(pk), pk, userProfileImage || root.store.generateIdenticon(pk))
+                return;
+            }
+>>>>>>> 27c4f26e7 (WIP refactor: removed dynamic scoping from shared module)
```
Comment (PascalPrecht) [inline ui/app/AppLayouts/Chat/views/ChatTextView.qml:111]: This is an unresolved merge conflict and needs to be solved.

File: ui/app/AppLayouts/Chat/views/ChatTextView.qml [hunk 5]
```diff
@@ -94,12 +94,21 @@ Item {
 //                return
 //            }
 
+<<<<<<< HEAD
```
Comment (Khushboo-dev-cpp) [inline ui/app/AppLayouts/Chat/views/ChatTextView.qml:97]: I think missed during merge conflict.

File: ui/app/AppLayouts/Chat/views/ChatTextView.qml (hunk 4 above)
Comment (Khushboo-dev-cpp) [inline ui/app/AppLayouts/Chat/views/ChatTextView.qml:111]: till here.

File: ui/app/AppLayouts/Profile/views/EnsView.qml [hunk 6]
```diff
@@ -318,25 +318,25 @@ Item {
...
-            toastMessage.iconColor = Style.current.primary
-            toastMessage.iconRotates = true
-            toastMessage.link = `${root.store.etherscanLink}/${txResult}`
-            toastMessage.open()
+            Global.toastMessage.title = qsTrId("ens-transaction-pending")
+            Global.toastMessage.source = Style.svg("loading")
+            Global.toastMessage.iconColor = Style.current.primary
+            Global.toastMessage.iconRotates = true
+            Global.toastMessage.link = `${root.store.etherscanLink}/${txResult}`
+            Global.toastMessage.open()
```
Comment (PascalPrecht) [inline ui/app/AppLayouts/Profile/views/EnsView.qml:326]: I might be wrong, but this looks like there can only be one `toastMessage` at a time.

I do agree that it makes sense for toastMessages to live in some global scope since they should be app section independent, but I also think that we should allow for multiple instances of this.

Does `toastMessage` have to be a prop of `Global`? And is it the one and only?

File: ui/imports/shared/popups/SendModal.qml [hunk 7]
```diff
@@ -33,33 +35,33 @@ ModalPopup {
...
-//                                                 txtAmount.selectedAsset.address,
-//                                                 txtAmount.selectedAmount,
-//                                                 gasSelector.selectedGasLimit,
-//                                                 gasSelector.eip1599Enabled ? "" : gasSelector.selectedGasPrice,
-//                                                 gasSelector.selectedTipLimit,
-//                                                 gasSelector.selectedOverallLimit,
-//                                                 transactionSigner.enteredPassword,
-//                                                 stack.uuid)
-//        }
+// Not Refactored Yet
```
Comment (Khushboo-dev-cpp) [inline ui/imports/shared/popups/SendModal.qml:38]: Was uncommenting this done on purpose?

File: ui/imports/shared/status/StatusChatInputTextFormationAction.qml [hunk 8]
```diff
@@ -3,12 +3,22 @@ import QtQuick.Controls 2.13
 
 Action {
     property string wrapper
+    property string selectedTextWithFormationChars: ""
```
Comment (Khushboo-dev-cpp) [inline ui/imports/shared/status/StatusChatInputTextFormationAction.qml:6]: move to the store? maybe?

File: ui/imports/shared/controls/AddressSourceSelector.qml [hunk 9]
```diff
@@ -40,9 +41,9 @@ Item {
         selectMenu.delegate: StatusMenuItemDelegate {
             statusPopupMenu: select
             action: StatusMenuItem {
-                text: root.sources[index].text
+                text: root.sources[currentIndex].text
```
Comment (PascalPrecht) [inline ui/imports/shared/controls/AddressSourceSelector.qml:44]: Uhm.. I think this is gonna error when `currentIndex == -1`.

Any reason we a) switched to `currentIndex` from `index` and b) this was initialized with `-1`?

File: ui/imports/shared/controls/ContactSelector.qml [hunk 10]
```diff
@@ -18,6 +18,7 @@ Item {
     id: root
     property var contacts
     property var selectedContact
+    property int currentIndex: -1
```
Comment (PascalPrecht) [inline ui/imports/shared/controls/ContactSelector.qml:21]: Since you're doing the same thing here, I'm sure there's a good reason for this. Does `currentIndex` have a special meaning in QML land?

**Review by Khushboo-dev-cpp — COMMENTED** [review]
Awesome!

Just left some comments/suggestions :)

File: ui/imports/shared/status/StatusChatInput.qml [hunk 11]
```diff
@@ -144,7 +145,7 @@ Rectangle {
 
     function checkTextInsert() {
         if (emojiSuggestions.visible) {
-            emojiSuggestions.addEmoji();
+            replaceWithEmoji(extrapolateCursorPosition(), emojiSuggestions.shortname, emojiSuggestions.unicode);
```
Comment (PascalPrecht) [inline ui/imports/shared/status/StatusChatInput.qml:148]: So this looks like a change in logic here...

Is this still part of dynamic scoping or does this actually do something else?
If this happens to be a fix or something, it'd be worth extracting this into its own commit.

File: ui/imports/shared/status/StatusStickerList.qml [hunk 12]
```diff
@@ -18,8 +20,8 @@ GridView {
     clip: true
     signal stickerClicked(string hash, int packId)
     delegate: Item {
-        width: stickerGrid.cellWidth
-        height: stickerGrid.cellHeight
```
Comment (PascalPrecht) [inline ui/imports/shared/status/StatusStickerList.qml:22]: This looks like this was referring to `cellWidth` and `cellHeight` from a different component.

Now this been changed to `root.cellWidth`/`root.cellHeight` respectively.
I don't see us passing those values to `StatusStickerList` in this PR.

So this means either: `width` and `height` of the item here was always broken, or these props potentially behave differently now.

What happened to the `stickerGrid` where those values came from?

(BTW: I see that `root` here has its own `cellWidth` and `cellheight`, but those haven't been introduced in this PR, so I want to make sure this change doesn't cause funny side effects)

**Review by PascalPrecht — COMMENTED** [review]
@alexandraB99 massive effort here!

Did a first review with some comments and questions. 
Let me know what you think!

File: ui/imports/shared/status/StatusChatInputTextFormationAction.qml (hunk 8 above)
Comment (alexandraB99) [reply ui/imports/shared/status/StatusChatInputTextFormationAction.qml:6]: moved!

File: ui/imports/shared/status/StatusStickerList.qml (hunk 12 above)
Comment (alexandraB99) [reply ui/imports/shared/status/StatusStickerList.qml:22]: So I see 3 usages of `StatusStickerList`. One is in `StatusStickerMarket.qml` where there is no such component 'id'ed' `stickerGrid` , the other one is in `StatusStickerPackClickPopup.qml` where the component with id `stickerGrid` is an `StatusStickerPackDetails` which is basically an `Item` and has no properties `cellWidth` and `cellheight` and the last place is in `StatusStickersPopup.qml` where the component that has id `stickerGrid` is the `StatusStickerList.qml` itself which is basically a `GridView` and has `cellWidth` and `cellheight` built-in properties set to 88 pixels each.  After this analysis I assumed that this binding wasn't making much sense and the at the end the desired binding was `root.cellWidth` / `root.cellHeight` instead of  `gridView.cellWidth` / `gridView.cellHeight`. Do you think I'm missing something?

File: ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml (hunk 2 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:53]: Actually I am not really sure what is/was the intention of this property, it is used only in `CompactMessageView.qml` in line 647, enabled property and it's actually set in both `ActivityCenterPopup.qml` and `ModalPopup.qml`. @PascalPrecht would you have any insights here?

File: ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml (hunk 1 above)
Comment (Khushboo-dev-cpp) [reply ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:81]: Hey Alex , I have a suggestion for this can we not have ENUMS defined for each of the sections in profile as well and have one single API.

Ex: appMain.changeAppSection(Constants.profile, Constants.ProfileSection.Browser) and in this way this can accommodate different things in the future for different modules? or just accept a second "params" in the same function.

Comment (alexandraB99) [reply ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml]: hey @Khushboo-dev-cpp updated, please have a look! I overlooked the places where the `changeProfileSection` was called and didn't realize that `changeAppSection` was also called before that and thus could be combined! thanks for pointing it out! :)

File: ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml (hunk 2 above)
Comment (PascalPrecht) [reply ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:53]: Hey friends, sorry for the late reply. Catching up on PR reviews...

So doing a quick search for `popupOpened` in the code base, we can see in the `CompactMessageView` that it's used as part of an expression to determine whether or not the hover handler for the message should be enabled.

I believe this was done because the UI was still signalling hover interaction in messages, even when a popup was overlaying the messages.
In other words, it's as if the cursor bleeds through the popup, causing the hover handler of messages to activate.

This is a hack and I assume this can be removed if we figure out why the popup doesn't prevent the underlying messages to receive mouse events.

File: ui/imports/shared/controls/AddressSourceSelector.qml (hunk 9 above)
Comment (PascalPrecht) [reply ui/imports/shared/controls/AddressSourceSelector.qml:43]: It looks like this is still here?

File: ui/imports/shared/status/StatusStickerList.qml (hunk 12 above)
Comment (PascalPrecht) [reply ui/imports/shared/status/StatusStickerList.qml:22]: This sounds reasonable, thanks for clarifying. If this works the same as before, this LGTM

**Review by PascalPrecht — COMMENTED** [review]
I think this is in good shape now, just that we're still using that `currentIndex` thingy

File: ui/imports/shared/controls/AddressSourceSelector.qml (hunk 9 above)
Comment (alexandraB99) [reply ui/imports/shared/controls/AddressSourceSelector.qml]: right, I missed that. Removed and rebased!

File: ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml (hunk 2 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:53]: yes, that's normal. what it's used to be done is to have a dummy disabled `MouseArea` inside the `ModalOverlay` to prevent background from consuming click events. Could you/is there already any issue for this?

File: ui/app/AppLayouts/Profile/views/EnsView.qml (hunk 6 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Profile/views/EnsView.qml:325]: Sounds good. Hopefully it's already filed in a github issue.

Comment (alexandraB99) [reply ui/app/AppLayouts/Profile/views/EnsView.qml:325]: When it's possible to open more than one toast messages, `Global.toastMessage` will become something like `Global.createToastMessage` or `Global.showToastMessage`, so in my opinion it's ok to keep `toastMessage` in `Global` for now. Let me know if you think differently!

File: ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml (hunk 2 above)
Comment (PascalPrecht) [reply ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:53]: This is the original issue https://github.com/status-im/status-desktop/issues/2187 and here's the PR that "fixed" this https://github.com/status-im/status-desktop/pull/2190

File: ui/app/AppLayouts/Profile/views/EnsView.qml (hunk 6 above)
Comment (PascalPrecht) [reply ui/app/AppLayouts/Profile/views/EnsView.qml:325]: > Sounds good. Hopefully it's already filed in a github issue.

There's not.

> When it's possible to open more than one toast messages, Global.toastMessage will become something like Global.createToastMessage or Global.showToastMessage, so in my opinion it's ok to keep 

Yea, I tend to agree. I don't think it make sense to have various multiple places where `ToastMessage` is declared because they most likely should behave the same, no matter from where they have been called.

Just important that different parts of the app get access to the service to spin them up.

So let's leave as is for now.

File: ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml (hunk 2 above)
Comment (alexandraB99) [reply ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:53]: I see. However this solves the issue only in the `HoverHandler` in `CompactMessage.qml`, should we work on a solution that's general to the Popups? That said, when a popup opens make sure that no events are consumed in the background?

Comment (PascalPrecht) [reply ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:53]: Oh yea absolutely. I rather see something more scalable and robust.
Would you mind creating an issue for this accordingly? Seems like you have a concrete idea of what that could look like.

Comment (jrainville) [reply ui/app/AppLayouts/Chat/popups/ActivityCenterPopup.qml:53]: Late reply, but I was the one that added that `popupOpened` property and Pascal got it right on the nose. It was indeed because there was a weird issue with the HoverHandler of CompactMessages that were getting interacted with when a popup was opened.
Feel free to remove `popupOpened` if you find a solution that works and is not as much of a hack.

**Review by Khushboo-dev-cpp — APPROVED** [review]

**Review by PascalPrecht — APPROVED** [review]
